File: android/core-domain/src/main/kotlin/com/crispy/tv/domain/home/HomeCatalogPlanner.kt
package com.crispy.tv.domain.home

import java.util.Locale

data class HomeCatalogItem(
    val id: String,
    val title: String,
    val posterUrl: String?,
    val backdropUrl: String?,
    val addonId: String,
    val type: String,
    val rating: String? = null,
    val year: String? = null,
    val description: String? = null,
)

enum class HomeCatalogSource(val key: String) {
    PERSONAL("personal"),
    PUBLIC("public");

    companion object {
        fun fromRaw(raw: String?): HomeCatalogSource? {
            return when (raw?.trim()?.lowercase(Locale.US)) {
                PERSONAL.key,
                "personal_home_feed",
                -> PERSONAL

                PUBLIC.key,
                "public_home_feed",
                -> PUBLIC

                else -> null
            }
        }
    }
}

enum class HomeCatalogPresentation(val key: String) {
    HERO("hero"),
    PILL("pill"),
    RAIL("rail");

    companion object {
        fun fromRaw(raw: String?): HomeCatalogPresentation {
            return when (raw?.trim()?.lowercase(Locale.US)) {
                HERO.key -> HERO
                PILL.key -> PILL
                else -> RAIL
            }
        }
    }
}

data class HomeCatalogList(
    val kind: String,
    val variantKey: String = DEFAULT_VARIANT_KEY,
    val source: HomeCatalogSource,
    val presentation: HomeCatalogPresentation = HomeCatalogPresentation.RAIL,
    val name: String = "",
    val heading: String = "",
    val title: String = "",
    val subtitle: String = "",
    val items: List<HomeCatalogItem>,
    val mediaTypes: Set<String> = emptySet(),
) {
    val catalogId: String
        get() = buildHomeCatalogId(source = source, kind = kind, variantKey = variantKey)

    val displayTitle: String
        get() = heading.ifBlank { title.ifBlank { name.ifBlank { kind } } }
}

data class HomeCatalogSnapshot(
    val profileId: String?,
    val lists: List<HomeCatalogList>,
    val statusMessage: String,
)

data class HomeCatalogHeroItem(
    val id: String,
    val title: String,
    val description: String,
    val rating: String?,
    val year: String? = null,
    val genres: List<String> = emptyList(),
    val backdropUrl: String,
    val addonId: String,
    val type: String,
)

data class HomeCatalogHeroResult(
    val items: List<HomeCatalogHeroItem> = emptyList(),
    val statusMessage: String = "",
)

data class HomeCatalogSection(
    val catalogId: String,
    val source: HomeCatalogSource,
    val presentation: HomeCatalogPresentation,
    val variantKey: String = DEFAULT_VARIANT_KEY,
    val name: String = "",
    val heading: String = "",
    val title: String = "",
    val subtitle: String = "",
) {
    val displayTitle: String
        get() = heading.ifBlank { title.ifBlank { name.ifBlank { catalogId } } }
}

data class HomeCatalogDiscoverRef(
    val section: HomeCatalogSection,
    val addonName: String,
    val genres: List<String> = emptyList(),
)

data class HomeCatalogFeedPlan(
    val heroResult: HomeCatalogHeroResult = HomeCatalogHeroResult(),
    val sections: List<HomeCatalogSection> = emptyList(),
    val sectionsStatusMessage: String = "",
)

data class HomeCatalogPageResult(
    val items: List<HomeCatalogItem> = emptyList(),
    val statusMessage: String = "",
    val attemptedUrls: List<String> = emptyList(),
)

data class HomeCatalogIdentifier(
    val source: HomeCatalogSource,
    val kind: String,
    val variantKey: String,
)

fun planHomeFeed(
    snapshot: HomeCatalogSnapshot,
    heroLimit: Int = 10,
    sectionLimit: Int = Int.MAX_VALUE,
): HomeCatalogFeedPlan {
    if (snapshot.lists.isEmpty()) {
        return HomeCatalogFeedPlan(
            heroResult = HomeCatalogHeroResult(statusMessage = snapshot.statusMessage),
            sections = emptyList(),
            sectionsStatusMessage = snapshot.statusMessage,
        )
    }

    return HomeCatalogFeedPlan(
        heroResult = buildHeroResult(snapshot, heroLimit),
        sections = buildHomeCatalogSections(snapshot.lists, sectionLimit),
        sectionsStatusMessage = "",
    )
}

fun planPersonalHomeFeed(
    snapshot: HomeCatalogSnapshot,
    heroLimit: Int = 10,
    sectionLimit: Int = Int.MAX_VALUE,
): HomeCatalogFeedPlan {
    return planHomeFeed(
        snapshot = snapshot,
        heroLimit = heroLimit,
        sectionLimit = sectionLimit,
    )
}

fun listDiscoverCatalogs(
    snapshot: HomeCatalogSnapshot,
    mediaType: String? = null,
    limit: Int = Int.MAX_VALUE,
): Pair<List<HomeCatalogDiscoverRef>, String> {
    val normalizedType = mediaType?.trim()?.lowercase(Locale.US)?.takeIf { it.isNotBlank() }
    if (normalizedType != null && normalizedType != "movie" && normalizedType != "series") {
        return emptyList<HomeCatalogDiscoverRef>() to "Unsupported media type: $mediaType"
    }

    if (snapshot.lists.isEmpty()) {
        return emptyList<HomeCatalogDiscoverRef>() to snapshot.statusMessage
    }

    val filteredLists =
        snapshot.lists.filter { list ->
            list.presentation == HomeCatalogPresentation.RAIL &&
                (normalizedType == null || list.supportsMediaType(normalizedType))
        }

    if (filteredLists.isEmpty()) {
        val suffix = if (normalizedType == null) "" else " for $normalizedType"
        return emptyList<HomeCatalogDiscoverRef>() to "No discover catalogs found$suffix."
    }

    val targetCount = limit.coerceAtLeast(1)
    val limitedLists = if (targetCount >= filteredLists.size) filteredLists else filteredLists.take(targetCount)
    return limitedLists.map { list ->
        HomeCatalogDiscoverRef(
            section = list.toSection(),
            addonName = "Supabase",
            genres = emptyList(),
        )
    } to ""
}

fun buildCatalogPage(
    snapshot: HomeCatalogSnapshot,
    sectionCatalogId: String,
    page: Int,
    pageSize: Int,
): HomeCatalogPageResult {
    val targetPage = page.coerceAtLeast(1)
    val targetSize = pageSize.coerceAtLeast(1)

    if (snapshot.lists.isEmpty()) {
        return HomeCatalogPageResult(
            items = emptyList(),
            statusMessage = snapshot.statusMessage,
            attemptedUrls = emptyList(),
        )
    }

    val identifier = parseHomeCatalogId(sectionCatalogId)
        ?: return HomeCatalogPageResult(
            items = emptyList(),
            statusMessage = "Catalog not found.",
            attemptedUrls = emptyList(),
        )

    val list =
        snapshot.lists.firstOrNull {
            it.source == identifier.source &&
                it.kind.equals(identifier.kind, ignoreCase = true) &&
                it.variantKey.equals(identifier.variantKey, ignoreCase = true)
        } ?: return HomeCatalogPageResult(
            items = emptyList(),
            statusMessage = "Catalog not found.",
            attemptedUrls = emptyList(),
        )

    val attempted =
        listOf(
            "supabase:${identifier.source.key}:${snapshot.profileId.orEmpty()}:${identifier.kind}:${identifier.variantKey}:page=$targetPage"
        )
    val allItems = list.items
    val startIndexLong = (targetPage.toLong() - 1L) * targetSize.toLong()
    if (startIndexLong >= allItems.size.toLong()) {
        return HomeCatalogPageResult(
            items = emptyList(),
            statusMessage = "No more items available.",
            attemptedUrls = attempted,
        )
    }

    val startIndex = startIndexLong.toInt()
    val endIndex = minOf(startIndex + targetSize, allItems.size)
    val items = if (startIndex < endIndex) allItems.subList(startIndex, endIndex) else emptyList()
    return HomeCatalogPageResult(
        items = items,
        statusMessage = when {
            items.isNotEmpty() -> ""
            targetPage <= 1 -> "No catalog items available."
            else -> "No more items available."
        },
        attemptedUrls = attempted,
    )
}

fun buildHomeCatalogId(
    source: HomeCatalogSource,
    kind: String,
    variantKey: String = DEFAULT_VARIANT_KEY,
): String {
    return "${source.key}:${kind.trim()}:${variantKey.trim().ifBlank { DEFAULT_VARIANT_KEY }}"
}

fun parseHomeCatalogId(catalogId: String): HomeCatalogIdentifier? {
    val trimmed = catalogId.trim()
    if (trimmed.isEmpty()) return null

    val parts = trimmed.split(':', limit = 3)
    if (parts.size < 3) return null

    val source = HomeCatalogSource.fromRaw(parts[0]) ?: return null
    val kind = parts[1].trim().takeIf { it.isNotEmpty() } ?: return null
    val variantKey = parts[2].trim().ifBlank { DEFAULT_VARIANT_KEY }
    return HomeCatalogIdentifier(
        source = source,
        kind = kind,
        variantKey = variantKey,
    )
}

fun resolveHomeCatalogSource(catalogId: String): HomeCatalogSource {
    return parseHomeCatalogId(catalogId)?.source ?: HomeCatalogSource.PERSONAL
}

private fun buildHeroResult(
    snapshot: HomeCatalogSnapshot,
    limit: Int,
): HomeCatalogHeroResult {
    val targetCount = limit.coerceAtLeast(1)
    val heroList = snapshot.lists.firstOrNull { it.presentation == HomeCatalogPresentation.HERO } ?: snapshot.lists.first()
    val fallbackDescription =
        heroList.subtitle.ifBlank {
            heroList.heading.ifBlank {
                heroList.title.ifBlank {
                    heroList.name.ifBlank { "Recommended for you." }
                }
            }
        }
    val heroItems =
        heroList.items
            .asSequence()
            .mapNotNull { item ->
                val backdrop = item.backdropUrl ?: item.posterUrl
                if (backdrop.isNullOrBlank()) return@mapNotNull null
                HomeCatalogHeroItem(
                    id = item.id,
                    title = item.title,
                    description = item.description ?: fallbackDescription,
                    rating = item.rating,
                    year = item.year,
                    genres = emptyList(),
                    backdropUrl = backdrop,
                    addonId = item.addonId,
                    type = item.type,
                )
            }
            .take(targetCount)
            .toList()

    if (heroItems.isEmpty()) {
        return HomeCatalogHeroResult(
            items = emptyList(),
            statusMessage = snapshot.statusMessage.ifBlank { "No featured items available." },
        )
    }

    return HomeCatalogHeroResult(items = heroItems, statusMessage = "")
}

private fun buildHomeCatalogSections(
    lists: List<HomeCatalogList>,
    limit: Int,
): List<HomeCatalogSection> {
    val targetCount = limit.coerceAtLeast(1)
    val filteredLists = lists.filterNot { it.presentation == HomeCatalogPresentation.HERO }
    if (filteredLists.isEmpty()) {
        return emptyList()
    }

    val limitedLists = if (targetCount >= filteredLists.size) filteredLists else filteredLists.take(targetCount)
    return limitedLists.map { it.toSection() }
}

private fun HomeCatalogList.toSection(): HomeCatalogSection {
    return HomeCatalogSection(
        catalogId = catalogId,
        source = source,
        presentation = presentation,
        variantKey = variantKey,
        name = name,
        heading = heading,
        title = title,
        subtitle = subtitle,
    )
}

private fun HomeCatalogList.supportsMediaType(mediaType: String): Boolean {
    return mediaTypes.any { it.equals(mediaType, ignoreCase = true) } ||
        items.any { it.type.equals(mediaType, ignoreCase = true) }
}

private const val DEFAULT_VARIANT_KEY = "default"

File: ios/ContractRunner/Sources/ContractRunner/HomeCatalogPlanner.swift
import Foundation

public struct HomeCatalogItem: Equatable {
    public let mediaKey: String
    public let title: String
    public let posterUrl: String?
    public let backdropUrl: String?
    public let addonId: String
    public let type: String
    public let rating: String?
    public let year: String?
    public let description: String?

    public init(
        mediaKey: String,
        title: String,
        posterUrl: String?,
        backdropUrl: String?,
        addonId: String,
        type: String,
        rating: String? = nil,
        year: String? = nil,
        description: String? = nil
    ) {
        self.mediaKey = mediaKey
        self.title = title
        self.posterUrl = posterUrl
        self.backdropUrl = backdropUrl
        self.addonId = addonId
        self.type = type
        self.rating = rating
        self.year = year
        self.description = description
    }
}

public enum HomeCatalogSource: String, Equatable {
    case personal = "personal"
    case publicFeed = "public"

    public static func fromRaw(_ raw: String?) -> HomeCatalogSource? {
        switch raw?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
        case HomeCatalogSource.personal.rawValue, "personal_home_feed":
            return .personal
        case HomeCatalogSource.publicFeed.rawValue, "public_home_feed":
            return .publicFeed
        default:
            return nil
        }
    }
}

public enum HomeCatalogPresentation: String, Equatable {
    case hero = "hero"
    case pill = "pill"
    case rail = "rail"

    public static func fromRaw(_ raw: String?) -> HomeCatalogPresentation {
        switch raw?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
        case HomeCatalogPresentation.hero.rawValue:
            return .hero
        case HomeCatalogPresentation.pill.rawValue:
            return .pill
        default:
            return .rail
        }
    }
}

public struct HomeCatalogList: Equatable {
    public let kind: String
    public let variantKey: String
    public let source: HomeCatalogSource
    public let presentation: HomeCatalogPresentation
    public let layout: String?
    public let name: String
    public let heading: String
    public let title: String
    public let subtitle: String
    public let items: [HomeCatalogItem]
    public let mediaTypes: Set<String>

    public init(
        kind: String,
        variantKey: String = "default",
        source: HomeCatalogSource,
        presentation: HomeCatalogPresentation = .rail,
        layout: String? = nil,
        name: String = "",
        heading: String = "",
        title: String = "",
        subtitle: String = "",
        items: [HomeCatalogItem],
        mediaTypes: Set<String> = []
    ) {
        self.kind = kind
        self.variantKey = variantKey
        self.source = source
        self.presentation = presentation
        self.layout = layout
        self.name = name
        self.heading = heading
        self.title = title
        self.subtitle = subtitle
        self.items = items
        self.mediaTypes = mediaTypes
    }

    public var catalogId: String {
        buildHomeCatalogId(source: source, kind: kind, variantKey: variantKey)
    }

    public var displayTitle: String {
        firstNonBlank(heading, title, name, kind) ?? kind
    }
}

public struct HomeCatalogSnapshot: Equatable {
    public let profileId: String?
    public let lists: [HomeCatalogList]
    public let statusMessage: String

    public init(profileId: String?, lists: [HomeCatalogList], statusMessage: String) {
        self.profileId = profileId
        self.lists = lists
        self.statusMessage = statusMessage
    }
}

public struct HomeCatalogHeroItem: Equatable {
    public let mediaKey: String
    public let title: String
    public let description: String
    public let rating: String?
    public let year: String?
    public let genres: [String]
    public let backdropUrl: String
    public let addonId: String
    public let type: String

    public init(
        mediaKey: String,
        title: String,
        description: String,
        rating: String?,
        year: String? = nil,
        genres: [String] = [],
        backdropUrl: String,
        addonId: String,
        type: String
    ) {
        self.mediaKey = mediaKey
        self.title = title
        self.description = description
        self.rating = rating
        self.year = year
        self.genres = genres
        self.backdropUrl = backdropUrl
        self.addonId = addonId
        self.type = type
    }
}

public struct HomeCatalogHeroResult: Equatable {
    public let items: [HomeCatalogHeroItem]
    public let statusMessage: String

    public init(items: [HomeCatalogHeroItem] = [], statusMessage: String = "") {
        self.items = items
        self.statusMessage = statusMessage
    }
}

public struct HomeCatalogSection: Equatable {
    public let catalogId: String
    public let source: HomeCatalogSource
    public let presentation: HomeCatalogPresentation
    public let layout: String?
    public let variantKey: String
    public let name: String
    public let heading: String
    public let title: String
    public let subtitle: String

    public init(
        catalogId: String,
        source: HomeCatalogSource,
        presentation: HomeCatalogPresentation,
        layout: String? = nil,
        variantKey: String = "default",
        name: String = "",
        heading: String = "",
        title: String = "",
        subtitle: String = ""
    ) {
        self.catalogId = catalogId
        self.source = source
        self.presentation = presentation
        self.layout = layout
        self.variantKey = variantKey
        self.name = name
        self.heading = heading
        self.title = title
        self.subtitle = subtitle
    }

    public var displayTitle: String {
        firstNonBlank(heading, title, name, catalogId) ?? catalogId
    }
}

public struct HomeCatalogDiscoverRef: Equatable {
    public let section: HomeCatalogSection
    public let addonName: String
    public let genres: [String]

    public init(section: HomeCatalogSection, addonName: String, genres: [String] = []) {
        self.section = section
        self.addonName = addonName
        self.genres = genres
    }
}

public struct HomeCatalogFeedPlan: Equatable {
    public let heroResult: HomeCatalogHeroResult
    public let sections: [HomeCatalogSection]
    public let sectionsStatusMessage: String

    public init(
        heroResult: HomeCatalogHeroResult = HomeCatalogHeroResult(),
        sections: [HomeCatalogSection] = [],
        sectionsStatusMessage: String = ""
    ) {
        self.heroResult = heroResult
        self.sections = sections
        self.sectionsStatusMessage = sectionsStatusMessage
    }
}

public struct HomeCatalogPageResult: Equatable {
    public let items: [HomeCatalogItem]
    public let statusMessage: String
    public let attemptedUrls: [String]

    public init(items: [HomeCatalogItem] = [], statusMessage: String = "", attemptedUrls: [String] = []) {
        self.items = items
        self.statusMessage = statusMessage
        self.attemptedUrls = attemptedUrls
    }
}

public struct HomeCatalogIdentifier: Equatable {
    public let source: HomeCatalogSource
    public let kind: String
    public let variantKey: String

    public init(source: HomeCatalogSource, kind: String, variantKey: String) {
        self.source = source
        self.kind = kind
        self.variantKey = variantKey
    }
}

public func planHomeFeed(
    snapshot: HomeCatalogSnapshot,
    heroLimit: Int = 10,
    sectionLimit: Int = .max
) -> HomeCatalogFeedPlan {
    guard !snapshot.lists.isEmpty else {
        return HomeCatalogFeedPlan(
            heroResult: HomeCatalogHeroResult(statusMessage: snapshot.statusMessage),
            sections: [],
            sectionsStatusMessage: snapshot.statusMessage
        )
    }

    return HomeCatalogFeedPlan(
        heroResult: buildHeroResult(snapshot: snapshot, limit: heroLimit),
        sections: buildHomeCatalogSections(lists: snapshot.lists, limit: sectionLimit),
        sectionsStatusMessage: ""
    )
}

public func planPersonalHomeFeed(
    snapshot: HomeCatalogSnapshot,
    heroLimit: Int = 10,
    sectionLimit: Int = .max
) -> HomeCatalogFeedPlan {
    planHomeFeed(snapshot: snapshot, heroLimit: heroLimit, sectionLimit: sectionLimit)
}

public func listDiscoverCatalogs(
    snapshot: HomeCatalogSnapshot,
    mediaType: String? = nil,
    limit: Int = .max
) -> ([HomeCatalogDiscoverRef], String) {
    let normalizedType = mediaType?
        .trimmingCharacters(in: .whitespacesAndNewlines)
        .lowercased()
        .nilIfBlank()

    if let normalizedType, normalizedType != "movie", normalizedType != "series" {
        return ([], "Unsupported media type: \(mediaType ?? "")")
    }

    guard !snapshot.lists.isEmpty else {
        return ([], snapshot.statusMessage)
    }

    let filteredLists = snapshot.lists.filter { list in
        list.presentation == .rail && (normalizedType == nil || list.supportsMediaType(normalizedType!))
    }

    guard !filteredLists.isEmpty else {
        let suffix = normalizedType.map { " for \($0)" } ?? ""
        return ([], "No discover catalogs found\(suffix).")
    }

    let targetCount = max(limit, 1)
    let limitedLists = targetCount >= filteredLists.count ? filteredLists : Array(filteredLists.prefix(targetCount))
    return (
        limitedLists.map { list in
            HomeCatalogDiscoverRef(section: list.toSection(), addonName: "Supabase", genres: [])
        },
        ""
    )
}

public func buildCatalogPage(
    snapshot: HomeCatalogSnapshot,
    sectionCatalogId: String,
    page: Int,
    pageSize: Int
) -> HomeCatalogPageResult {
    let targetPage = max(page, 1)
    let targetSize = max(pageSize, 1)

    guard !snapshot.lists.isEmpty else {
        return HomeCatalogPageResult(items: [], statusMessage: snapshot.statusMessage, attemptedUrls: [])
    }

    guard let identifier = parseHomeCatalogId(sectionCatalogId) else {
        return HomeCatalogPageResult(items: [], statusMessage: "Catalog not found.", attemptedUrls: [])
    }

    guard let list = snapshot.lists.first(where: {
        $0.source == identifier.source &&
            $0.kind.caseInsensitiveCompare(identifier.kind) == .orderedSame &&
            $0.variantKey.caseInsensitiveCompare(identifier.variantKey) == .orderedSame
    }) else {
        return HomeCatalogPageResult(items: [], statusMessage: "Catalog not found.", attemptedUrls: [])
    }

    let attempted = [
        "supabase:\(identifier.source.rawValue):\(snapshot.profileId ?? ""):\(identifier.kind):\(identifier.variantKey):page=\(targetPage)"
    ]
    let startIndexLong = Int64(targetPage - 1) * Int64(targetSize)
    guard startIndexLong < Int64(list.items.count) else {
        return HomeCatalogPageResult(items: [], statusMessage: "No more items available.", attemptedUrls: attempted)
    }

    let startIndex = Int(startIndexLong)
    let endIndex = min(startIndex + targetSize, list.items.count)
    let items = startIndex < endIndex ? Array(list.items[startIndex..<endIndex]) : []
    let statusMessage: String
    if !items.isEmpty {
        statusMessage = ""
    } else if targetPage <= 1 {
        statusMessage = "No catalog items available."
    } else {
        statusMessage = "No more items available."
    }

    return HomeCatalogPageResult(items: items, statusMessage: statusMessage, attemptedUrls: attempted)
}

public func buildHomeCatalogId(
    source: HomeCatalogSource,
    kind: String,
    variantKey: String = "default"
) -> String {
    let normalizedVariantKey = variantKey.trimmingCharacters(in: .whitespacesAndNewlines).nilIfBlank() ?? defaultVariantKey
    return "\(source.rawValue):\(kind.trimmingCharacters(in: .whitespacesAndNewlines)):\(normalizedVariantKey)"
}

public func parseHomeCatalogId(_ catalogId: String) -> HomeCatalogIdentifier? {
    let trimmed = catalogId.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else {
        return nil
    }

    let parts = trimmed.split(separator: ":", maxSplits: 2, omittingEmptySubsequences: false).map(String.init)
    guard parts.count == 3 else {
        return nil
    }
    guard let source = HomeCatalogSource.fromRaw(parts[0]) else {
        return nil
    }
    let kind = parts[1].trimmingCharacters(in: .whitespacesAndNewlines)
    guard !kind.isEmpty else {
        return nil
    }
    let variantKey = parts[2].trimmingCharacters(in: .whitespacesAndNewlines).nilIfBlank() ?? defaultVariantKey
    return HomeCatalogIdentifier(source: source, kind: kind, variantKey: variantKey)
}

public func resolveHomeCatalogSource(catalogId: String) -> HomeCatalogSource {
    parseHomeCatalogId(catalogId)?.source ?? .personal
}

private func buildHeroResult(snapshot: HomeCatalogSnapshot, limit: Int) -> HomeCatalogHeroResult {
    let targetCount = max(limit, 1)
    let heroList = snapshot.lists.first(where: { $0.presentation == .hero }) ?? snapshot.lists[0]
    let fallbackDescription = firstNonBlank(
        heroList.subtitle,
        heroList.heading,
        heroList.title,
        heroList.name,
        "Recommended for you."
    ) ?? "Recommended for you."
    let heroItems = heroList.items.compactMap { item -> HomeCatalogHeroItem? in
        let backdrop = item.backdropUrl ?? item.posterUrl
        guard let backdrop, !backdrop.isEmpty else {
            return nil
        }
        return HomeCatalogHeroItem(
            mediaKey: item.mediaKey,
            title: item.title,
            description: item.description ?? fallbackDescription,
            rating: item.rating,
            year: item.year,
            genres: [],
            backdropUrl: backdrop,
            addonId: item.addonId,
            type: item.type
        )
    }
    let limitedItems = Array(heroItems.prefix(targetCount))

    if limitedItems.isEmpty {
        return HomeCatalogHeroResult(
            items: [],
            statusMessage: snapshot.statusMessage.nilIfBlank() ?? "No featured items available."
        )
    }
    return HomeCatalogHeroResult(items: limitedItems, statusMessage: "")
}

private func buildHomeCatalogSections(lists: [HomeCatalogList], limit: Int) -> [HomeCatalogSection] {
    let filteredLists = lists.filter { $0.presentation != .hero }
    guard !filteredLists.isEmpty else {
        return []
    }
    let targetCount = max(limit, 1)
    let limitedLists = targetCount >= filteredLists.count ? filteredLists : Array(filteredLists.prefix(targetCount))
    return limitedLists.map { $0.toSection() }
}

private extension HomeCatalogList {
    func toSection() -> HomeCatalogSection {
        HomeCatalogSection(
            catalogId: catalogId,
            source: source,
            presentation: presentation,
            layout: layout,
            variantKey: variantKey,
            name: name,
            heading: heading,
            title: title,
            subtitle: subtitle
        )
    }

    func supportsMediaType(_ mediaType: String) -> Bool {
        mediaTypes.contains(where: { $0.caseInsensitiveCompare(mediaType) == .orderedSame }) ||
            items.contains(where: { $0.type.caseInsensitiveCompare(mediaType) == .orderedSame })
    }
}

private extension String {
    func nilIfBlank() -> String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}

private func firstNonBlank(_ values: String?...) -> String? {
    for value in values {
        if let normalized = value?.trimmingCharacters(in: .whitespacesAndNewlines), !normalized.isEmpty {
            return normalized
        }
    }
    return nil
}

private let defaultVariantKey = "default"
